tests/Microsoft.Identity.Test.Unit/CoreTests/RegionDiscoveryProviderTests.cs

```diff
@@ -85,6 +85,43 @@ public async Task SuccessfulResponseFromLocalImdsAsync()
             ValidateInstanceMetadata(regionalMetadata);
         }
 
+        [TestMethod]
+        public void MultiThreadSuccessfulResponseFromLocalImds_HasOnlyOneImdsCall()
+        {
+            const int MaxThreadCount = 5;
+            // add the mock response only once and call it 5 times on multiple threads
+            // if the http mock is called more than once, it will fail in dispose as queue will be non-empty
+            AddMockedResponse(MockHelpers.CreateSuccessResponseMessage(TestConstants.Region));
+            SemaphoreSlim semaphore = new SemaphoreSlim(0);
+            int threadCount = MaxThreadCount;
+            var result = Parallel.For(0, MaxThreadCount, async (i) =>
+            {
+                try
+                {
+
+                    _testRequestContext.ServiceBundle.Config.AzureRegion =
+                        ConfidentialClientApplication.AttemptRegionDiscovery;
+
+                    InstanceDiscoveryMetadataEntry regionalMetadata = await _regionDiscoveryProvider.GetMetadataAsync(
+                        new Uri("https://login.microsoftonline.com/common/"), _testRequestContext).ConfigureAwait(false);
+
+                    ValidateInstanceMetadata(regionalMetadata);
+                    Interlocked.Decrement(ref threadCount);
```

Comment:
I think this needs to be in a `finally` statement.

I removed the semaphores from the RegionManager code and this test hung up. There were exception in all threads because regionalMetadata was null, for some reason. So threadCount was never decremented.

Reply:
Good catch